### ISSUE TITLE
Get rid of File

### ISSUE THREAD

**Issue by milliams** [ganga-devs/ganga#412]
Currently in a number of Ganga APIs we have the option of using the `File` type. This is completely separate and unrelated to `IGangaFile` (and therefore `LocalFile`, `DiracFile` etc.). I believe that historically (before we had an inputfiles/outputfiles system) it was used to signal that, for example, the file should be copied into the user's sandbox.

It is currently usable in a number of places, for example in `Executable.exe` and `CustomChecker.module`. The problem is that is used very inconsistently. The API they seem to _want_ to provide is the ability to seamlessly do either of the following:

``` python
j.application.exe = 'myscript.sh'
j.application.exe = File('myscript.sh')
```

But there is a fundamental problem that `File` is a `GangaObject` and therefore needs to be recorded as a `ComponentItem` in the schema but a string is not and must be recorded as a `SimpleItem`. This means that we've ended up in a bunch of different 'solutions' in the code, the main two that I've seen are:

``` python
ComponentItem(category='files')  # e.g. CustomChecker.module
SimpleItem(defvalue='echo', typelist=[str, File])  # e.g. Executable.exe
```

Due to Ganga's currently lax and inconsistent type checking they are both able to accept both `str` and `File` objects via assignment but what happens afterwards only seems to work (where it _does_ work) by chance.
# Serialisation

Consider the following 4 variants:

``` python
box.add(Executable(exe='~/mychecker.py'), 'e')  # 1. SimpleItem str
box.add(Executable(exe=File('~/mychecker.py')), 'e2')  # 2. SimpleItem File
box.add(CustomChecker(module='~/mychecker.py'), 'cc')  # 3. ComponentItem str
box.add(CustomChecker(module=File('~/mychecker.py')), 'cc2')  # 4. ComponentItem File
```

When they are serialised they end up with very different XML outputs (snipped):

``` xml
1.
<attribute name="exe"> 
  <value>'~/mychecker.py'</value>
</attribute>

2.
<attribute name="exe"> 
  <value>File(name='/afs/cern.ch/user/m/mwilliam/mychecker.py',subdir='.')</value>
</attribute>

3.
<attribute name="module">
  <value>'~/mychecker.py'</value>
</attribute>

4.
<attribute name="module">
  <class name="File" version="1.1" category="files">
    <attribute name="name"> 
      <value>'/afs/cern.ch/user/m/mwilliam/mychecker.py'</value>
    </attribute>
    <attribute name="subdir"> 
      <value>'.'</value>
    </attribute>
  </class>
</attribute>
```

`SimpleItem`/`str` (1) and `ComponentItem`/`File` (4) both work as expected and round-trip correctly but 2 and 3 do not. 2 streams out 'successfully' as `SimpleItem`s use the object's `repr`. 3 streams out 'successfully' due to an unassuming (and unexplained) `if isType(s, str)` in `acceptOptional` (added by @rob-c in 717775211e84ea). The latter is not present in `VPrinter` so that a plain `CustomChecker(module='~/mychecker.py')` will raise an exception.

As for reading (after a Ganga restart), `box['e']` works and `exe` is set to a string, `box['e2']` works with `exe` set to `File` (which works due to the stored XML data being passed through `eval`). `box['cc']` fails due to the inconsistent types and `box['cc2']` works with `module` being `File`.
# Usage

`File` has two public attributes, `name` and `subdir`. On the whole, only `name` is ever used (and people will just put the full path in `name`). For example, `Executable`, `CustomChecker`, `CustomMerger`, `Gaudi` and co., `Bender`, and `Athena` and co. just use `name` and never `subdir`. Only `Root` and `GaudiPython` use `subdir`.
# Partial proposal

I suggest that we find a way to get rid of `File` as it does not add any _functionality_ over a simple string. There is an open question about a safe transition mechanism (both from the point-of-view of the XML repository and also and user scripts) about which I am open to discussion.

**Comment by rob-c** [comment]
I would like to agree here and I really, really want to. But this is a huge task imo and we __still__ don't have full feature compatibility between LocalFile and File. And by that I mean some ```remoteDir``` attribute which works. This is something we still need in IGangaFile as well.

I would really like to see this go but I wonder if this is the equivalent of a Ganga 6.3 (or higher) task as it's unfortunately invasive.

**Comment by milliams** [comment]
`LocalFile` and `File` are completely different things and they're never going to have feature compatibility. Are you suggesting that `File` needs a `remoteDir` or that `LocalFile` does?

**Comment by rob-c** [comment]
#344 We need the ability to put files into a directory on the worker node once we land there. This is something holding some LHCb users back on 6.0.44 and yes they could modify their payload to move the file but they often can't.
Once ```LocalFile``` can be used as a drop in replacement for ```File``` then I say we try and do this wherever possible and in the process of doing that remove it where it's not needed but I think we should be expanding the functionality rather than reducing it.

**Comment by milliams** [comment]
I agree that something like that would be useful for `IGangaFile` but I guess I'm not understanding how that relates to `File`. `File` offers nothing of the sort at the moment, it's simply a holder for `name` and `subdir`. Where's the code which uses `File` to do the remote directory handling?

**Comment by rob-c** [comment]
The code doesn't since 6.1.
That was the point of 6.1, to move away from submitting ```File``` objects in the ```sandbox``` to submitting ```IGangaFile``` objects in the ```inputfiles```. The ```File``` object was storing the name of the file and the directory that the file should be placed in on the worker node.

**Comment by rob-c** [comment]
Yes the file object then could have been replaced with a ```tuple``` or some equivalent but amongst other things the type was useful to keep track of objects.
There are many places where the File can be dropped for a simple string/tuple but a lot of these places are places such as ```.exe``` where we as developers would stick nothing but a string but then I  feel someone will want to execute a program from a subdirectory one day as crazy as that may be.

**Comment by drmarkwslater** [comment]
@rob-c So if I'm understanding correctly, there are two outstanding issues:

1. Some users are requesting a system of making sure files go in a particular sub folder on the WN
2. We may want to support a user running an Executable from a sub folder.

For 1) I'd say that this is something that needs to be added to `IGangaFile`  and if it's holding some users back from using 6.1 then we should probably get that added ASAP. For 2), `File` doesn't provide that anyway and certainly for `Executable` jobs I can't see any way of really doing that short of explicitly adding it into the application (which I wouldn't do anyway unless it was specifically requested 😄  ). So aren't both of these orthogonal to removing `File` as it stands in 6.1? It's not used in Sandboxes anymore and the only other place would be in applications that just use it as a string anyway. Or am I missing something?

Now having said all that, what the best thing to do as regards phasing it out and not annoying lots of people is a much harder question...

**Comment by egede** [comment]
@milliams I do not conceptually see that the `File` and `IGangaFile` are different. In fact I believe that every place that currently accepts a `File` object could accept any `IGangaFile` object. Would it not be perfectly reasonable if I in the `Executable` application could give a `DiracFile` as the executable?

**Comment by drmarkwslater** [comment]
I agree that would be a nice feature, but when we were discussing it I wasn't sure how hard that would be to implement. If we can do it though, I'd be all for it :smile:

**Comment by egede** [comment]
>  agree that would be a nice feature, but when we were discussing it I wasn't sure how hard that would be to implement. If we can do it though, I'd be all for it.

I think it could simply use the same code injection code as we do for the input files. There is something funny at the moment where some rather unclear code decides that if you do `application.exe=echo` it will not actually copy the `echo` executable but I think we could in fact get rid of that.

**Comment by drmarkwslater** [comment]
That's fine - my possible issue was providing a mechanism for an Application (or any other plugin for that matter) to make the submission process aware of these extra files, i.e. at present only `inputfiles` is checked so we'd need an agreed API for additional components to be checked at submit time. Maybe we could get away with just adding an overloadable function to IApplication that returns any additional files to be considered for input?

**Comment by egede** [comment]
@drmarkwslater Yes, we should certainly not just add them to the `inputfiles`. This is what we in the past have done for output extracted from LHCb options files and it causes lots of trouble when object are subsequently copied. As the information is transient, it should not be too much of an issue to implement this.

**Comment by rob-c** [comment]
@drmarkwslater I think we would benefit from a common function in the prepared application interface (or elsewhere) to get the files intended for the inputfiles and maybe the inputdata based upon what the application knows/wants. This is currently all done in each RTHandler and is a little fragile.
Doing would mean that the RTHandlers would reduce in complexity by quite a bit and there would be a clearer line between the prepare application and prepare job steps imo.

**Comment by rob-c** [comment]
@egede This unfortunately is what we're still doing in the RTHandlers as far as I can tell and what I've ended up doing in the LSST plugin, although I've a sort of interface between the 2 layers even if it gets confused a bit.

**Comment by drmarkwslater** [comment]
@rob-c I agree - it's a fairly common thing of Applications needing to send additional files over so providing an easy interface for this that means the RTHandlers can be simplified is a very good idea imo :smile:

**Comment by rob-c** [comment]
This is another issue which would flow a lot quicker with a blackboard of some sort and a round table I think.
